### ISSUE TITLE
@ApiResponse cannot infer parametrised type with Java Generics 

### ISSUE THREAD

**Issue by kliarist** [swagger-api/swagger-core#4355]
I have the following code snippet

```
public class ApiStandardResponse<T> {
    T data;
}
```

and use that to annotate my endpoint 

```
@Operation(
   ...
      responses = {
          @ApiResponse(
              responseCode = "200",
              description = "Success",
              content = {@Content(
                  mediaType = "application/json",
                  schema = @Schema(implementation = ApiStandardResponse.class)
              )}
          )
      }
   ...
  )
```

It turns out that there is no way to specify the parametrised type and thus swagger ui cannot infer it making it impossible for contract clients to generate the appropriate stubs 

![image](https://user-images.githubusercontent.com/1371419/214840426-b2aa0f94-60c2-4a93-b998-284b356c68ec.png)

Is there an elegant way to specify the parametrised type somehow without using workarounds such as subclasses?

**Comment by FranH20** [comment]
I have the same question.

**Comment by FranH20** [comment]
Hello,

I had the same issue, so I found a work-around that I would like to share.

I had a class with generic type named CommonResponse<T>, that I wanted to use as the response type for swagger.
I defined an extra class, named UserResponse that extends CommonResponse<User> and provides a getter for User instance variable, and it seems to work as expected now.

Good luck!

**Comment by kliarist** [comment]
@FranH20 Yeah I ended up doing the same when faced with this issue . It is not ideal as you end up having many different subclasses (depending how many different dto-response object you have), but works.

**Comment by Linkk2000** [comment]
Checking in on this issue in 2025, as I'm encountering the same problem.
Like others, I'm looking for a solution that avoids generating potentially unnecessary subclasses.
Has there been any progress on alternative approaches? I've considered possibilities like runtime bytecode generation (e.g., in Java), but I'm wondering if there's a more officially recommended or cleaner pattern to solve this now?
Any updates or suggestions would be appreciated.

**Comment by bsayli** [comment]
Hi all — I’ve hit the same limitation (where `@Schema(implementation = ApiStandardResponse.class)` can’t express the type parameter) and ended up with a small, production-ready pattern that preserves the generic envelope while generating clean, type-safe clients.

---

### 🔊 What worked for us (no subclass explosion)

**1) Keep the generic envelope on the server**

```java
@GetMapping("/v1/customers/{id}")
public ResponseEntity<ServiceResponse<CustomerDto>> getCustomer(@PathVariable int id) {
  return ResponseEntity.ok(ServiceResponse.ok(service.getCustomer(id)));
}
```

**2) At OpenAPI publish time, inject a composed schema + vendor hints**

```yaml
components:
  schemas:
    ServiceResponseCustomerDto:
      allOf:
        - $ref: "#/components/schemas/ServiceResponse"
        - type: object
          properties:
            data:
              $ref: "#/components/schemas/CustomerDto"
      x-api-wrapper: true
      x-api-wrapper-datatype: CustomerDto
      # optional when nested:
      # x-data-container: Page
      # x-data-item: CustomerDto
```

A small Springdoc customizer inspects controller return types like `ServiceResponse<T>` or `ServiceResponse<Page<T>>` and automatically adds the composed schema with these extensions.

**3) On the client, a tiny Mustache overlay emits thin wrappers**

```mustache
{{! api_wrapper.mustache }}
public class {{classname}}
  extends com.example.client.ServiceClientResponse<
    {{#vendorExtensions.x-data-container}}
      {{vendorExtensions.x-data-container}}<{{vendorExtensions.x-data-item}}>
    {{/vendorExtensions.x-data-container}}
    {{^vendorExtensions.x-data-container}}
      {{vendorExtensions.x-api-wrapper-datatype}}
    {{/vendorExtensions.x-data-container}}
  > {}
```

Resulting Java:

```java
public class ServiceResponseCustomerDto
    extends ServiceClientResponse<CustomerDto> {}

public class ServiceResponsePageCustomerDto
    extends ServiceClientResponse<Page<CustomerDto>> {}
```

---

### 🔹 Why this helps

* No more `ResponseLessThanByteStringGreaterThan` name mangling.
* No subclass proliferation in server code.
* DRY envelope: `{ data, meta }` lives once in a generic base.
* Clean compile-time typing (`getData()` returns `T`).
* Works with Spring Boot 3.4 + Springdoc 2.8 + OpenAPI Generator 7.x.

---

### 🔹 Possible action for swagger-core

If swagger-core provided a way to express parameterized `implementation` (e.g., an annotation attribute or recognized `x-` extension for type parameters), vendor hints like these wouldn’t be necessary.

Even minimal support for extensions such as:

```yaml
x-api-wrapper: true
x-api-wrapper-datatype: CustomerDto
```

would enable fully type-safe schema definitions without subclass bloat.

Happy to share more details or contribute docs if this direction sounds reasonable.

---

Reference implementation (runnable):
🔗 [blueprint-platform/openapi-generics](https://github.com/blueprint-platform/openapi-generics)
Write-up (why + how):
📖 [We Made OpenAPI Generator Think in Generics](https://medium.com/@baris.sayli/type-safe-generic-api-responses-with-spring-boot-3-4-openapi-generator-and-custom-templates-ccd93405fb04)
Adoption guide de eklemek istersen:
📘 [Adoption Guide](https://blueprint-platform.github.io/openapi-generics/)
